### PR TITLE
Allow admin to decline a user's "This is My Organization" request without deleting the user

### PR DESCRIPTION
**Pull request by marianmosley**

### CODE REVIEW

**Comment by tansaku** [comment]
Thanks @marianmosley - so this is for https://www.pivotaltracker.com/story/show/82642350 right?

This is a great first pass!

I notice you've introduced a new end point on user_reports:

`user_reports#remove_pending_org_from_user`

which may be the right way to go here, but it sort of breaks the RESTful model that I think we should be following whereby we try to use only the existing RESTful verbs of 'update', 'destroy' etc.

Now maybe the setup we have for the user_reports controller is such that that is too painfully hard to support, but is it worth having a think about whether we can pull this logic into the 'update' method?  I know we've got a service in there, but as you can see from the comment in the file, I'd love to have 'update' support generic updating of the model, and then the business logic pulled out into services or object lifecycle hooks.  Does that make any sense?

Slightly less important, when I look at:

```ruby
And /"(.*)"'s request for "(.*)" should (not )?be persisted/ do |email, org, negative|
  user = User.find_by_email(email)
  org = Organisation.find_by_name(org)
  expectation = negative ? :not_to : :to
  expect(user.pending_organisation).send(expectation, eq(org))
end
```

I'm struck (not for the first time with this pattern), about how supporting the negation in the cucumber cleanly ends up making the rspec in the step less readable.  I don't know if there is a better solution for this that keeps readability in both contexts ...

**Comment by marianmosley** [comment]
Yes - I think adding the functionality to the update method would work.  I will look at that.

What do you think of this less elegant but perhaps more readable step definition?
```
And /"(.*)"'s request for "(.*)" (should|should not) be persisted/ do |email, org, expectation|
  user = User.find_by_email(email)
  org = Organisation.find_by_name(org)
  case expectation
  when 'should'
    expect(user.pending_organisation).to eq(org)
  when 'should not'
    expect(user.pending_organisation).to be_nil
  end
end
```

**Comment by tansaku** [comment]
thanks @marianmosley - I think that is easier to read - I wonder what others think @johnnymo87 @mtc2013 @decareano @ZmagoD ?

**Comment by johnnymo87** [comment]
I'm fine with either.

**Comment by mtc2013** [comment]
Either version works for me @tansaku

**Comment by decareano** [comment]
marian's code is very easy to read. Code with ternary operator is more compact (3 less lines) but more difficult to read. either will do for me.

**Comment by ZmagoD** [comment]
I agree with that marian's code is easier to read. Either will do for me.

**Comment by mtc2013** [comment]
@tansaku there have been some updates to this pr.

**Comment by tansaku** [comment]
thanks @marianmosley 

overall great - could do a oneliner in user decliner

```ruby
    if !is_current_user_superadmin?
      listener.authorization_failure_for_update
    end
```
to
```ruby
      listener.authorization_failure_for_update unless current_user.superadmin?
```

I still wonder if rather than services like this we should prefer to be updating models directly through the relevant controller with appropriate lifecycle hooks ...

UserOrganisationDecliner is checking for permissions and then updating the user - would it not be cleaner to do this through the user controller?  I'm not suggesting you undertake any more refactoring here.  I'm just thinking out loud about the user of services vs RESTful pathways for updating objects ...

it's interesting to look at the update code we have in UsersController:

```ruby
  def update
    usr = User.find_by_id UserParams.build(params).fetch(:id)
    UserOrganisationClaimer.new(self, usr, usr).call(UserParams.build(params).fetch(:pending_organisation_id))
  end
```

**Comment by mtc2013** [comment]
@tansaku I've pushed a commit to make that into a one liner.

File: app/services/user_organisation_decliner.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  def call
+    error_message_if_not_superadmin
+    remove_pending_org_from_user
+  end
+
+  attr_reader :listener, :user, :current_user
+
+  private
+
+  def is_current_user_superadmin?
```
Comment (tansaku) [inline app/services/user_organisation_decliner.rb]: is this method really necessary - feels like it's just repeating the already nicely descrptive `current_user.superadmin?`

**Comment by tansaku** [comment]
thanks @mtc2013 - I made one other minor code comment and I guess we should just pull this in, but I'd love to have input from @johnnymo87 @marianmosley @decareano @ZmagoD on what they think about the services vs RESTful model updates approach.  I don't want to block this functionality change with that discussion, but would love to get everyone's input.  

It seems to me that services should be for things without state such as sending emails etc. and that we should try to move towards a setup where model modifications are coordinated through RESTful update methods?  What does everything think?  Not a bit deal either way?  Pros and cons?

File: app/services/user_organisation_decliner.rb (hunk 1 above)
Comment (mtc2013) [inline app/services/user_organisation_decliner.rb]: It arguably isolates us from devise, I suppose.  It's the same style as in another service we have.

Comment (mtc2013) [inline app/services/user_organisation_decliner.rb]: @tansaku actually, since we inject current_user into the new, we don't really need this, I suppose.

**Pull request merged**